service/src/main/java/bio/terra/profile/db/ProfileDao.java

```diff
@@ -76,21 +80,30 @@ public BillingProfile createBillingProfile(
             .addValue("created_by", user.getEmail());
 
     var keyHolder = new DaoKeyHolder();
-    jdbcTemplate.update(sql, params, keyHolder);
-
-    return new BillingProfile(
-        profile.id(),
-        profile.displayName(),
-        profile.description(),
-        profile.biller(),
-        profile.cloudPlatform(),
-        profile.billingAccountId(),
-        profile.tenantId(),
-        profile.subscriptionId(),
-        profile.managedResourceGroupId(),
-        keyHolder.getInstant("created_date"),
-        keyHolder.getInstant("last_modified"),
-        keyHolder.getString("created_by"));
+    try {
+      jdbcTemplate.update(sql, params, keyHolder);
+
+      return new BillingProfile(
+          profile.id(),
+          profile.displayName(),
+          profile.description(),
+          profile.biller(),
+          profile.cloudPlatform(),
+          profile.billingAccountId(),
+          profile.tenantId(),
+          profile.subscriptionId(),
+          profile.managedResourceGroupId(),
+          keyHolder.getInstant("created_date"),
+          keyHolder.getInstant("last_modified"),
+          keyHolder.getString("created_by"));
+    } catch (DuplicateKeyException ex) {
```

Comment:
I added this (and the dedicated exception) so that we could distinguish between a `DuplicateKeyException` because the billing profile id is already in use and one where the managed app is already in use. That said, I don't love having to make the determination based on the contents of the exception message. It will be pretty stable as the substring we're searching for is the name of the uniqueness constraint which is theoretically pretty stable. Still, if anyone has a better idea of how to do this, I'm all ears.